keps/0031-namespace-management.md

```diff
@@ -0,0 +1,152 @@
+
+### The Many Meanings of a Namespace
+
+Challenages to auto-provision of namespaces are complicated by the many reasons / values for creating them which includes:
+
+1. Logical separation of work (separating 1000s of pods into 10s of buckets for management)
+1. Quota management
+1. Security separation
+1. Metadata (where labels and annotations provide meaning and value to controllers managing resources in a namespace)
+
+Based on the last 3 reasons, it is common for an administrator to create namespaces to manage the other concerns associated with a namespace.  It would be challeging to manage a cluster where dozen or more namespaces are adhoc being created (or deleted) which may require a host of other configurations around quota and security.
+
+## Proposals
+
+Prior to making KEP `provisional` we will chose a namespace creation proposal to move forward with and move the 2nd proposal to alternative.  For now, we have 2 competing namespace creation proposals.  
+In addition there is a proposal for the use of multiple namespaces.
+
+### Proposal 1: No namespace creation
+
+All namespace creation is managed outside KUDO. Users such as D2iQ would need to find or build automation for the creation of their namespaces. This REQUIRES that KUDO honor the deployment of resources to the "default" namespace, or to the namespace defined by `--namespace` during install.  
+
+KUDO should further require that NO namespaces be allow in manifests for resources deployed during "install".  The concept of defining manifest which land in multiple namespaces is defined in a latter proposal and is a separate concern. Without multi-namespace support, all resources are install to one namespace and namespace must be created prior to installation and that namespace must be specified during installation.
+
+### Proposal 2: Namespace Creation
+
+KUDO should support the creation of a namespace if it does not exist.  Further KUDO will add a configuration for namespace such that at time of namespace creation it is created with the metadata provided. To this end a manifest file for namespace is to be supported in the template folder:
+
+```
+apiVersion: v1
+kind: Namespace
+metadata:
+  labels:
+    ca.istio.io/override: "true"
+    istio-injection: enabled
+    katib-metricscollector-injection: enabled
+  name: {{ .Namespace }}
+```
+
+The template supports the implicit `.Namespace` parameter (which already exists).  The `operator.yaml` file is to will be extended to support `namespaceManifest`.
+
+```
+# operator.yaml
+name: "kubeflow"
+operatorVersion: "0.2.0"
+kudoVersion: "0.10.1"
+kubernetesVersion: 1.15.0
+appVersion: 1.0.0
+namespaceManifest: templates/namespace.yaml
+```
+
+The follow rules apply:
+
+1. Missing `namespaceManifest` means that there is no metadata (creation still applies if warranted)
+1. Missing `name: {{ .Namespace }}` is a failure (and will be validated by package verifier)
```

Comment:
I assume this means exactly `name: {{ .Namespace }}` correct? i.e. `name: {{ .Namespace }}-storage` would be an error as well?

Reply:
great point... we need to include this restriction as well... I think this is why I was hesitate to allow templating... it should just be a namespace manifest with no name that will be named by kudo. (via `kudo install --namespace` or default)

Reply:
with more thought on this... I'm of the opinion that the best solution is to NOT allow name to be specified... the name will be provided and controlled by kudo.  This manifest file is a namespace manifest file and is only useful if creating a namespace... in which case kudo will provide the name.